README.md

```diff
@@ -1,161 +1,720 @@
 
-A relation is a set of tuples, where each component of the tuple is called an attribute, and can be referenced by an attribute name.
+All such representations MUST form a partial order, such that subsequent epochs are represented using subsequent timestamps. For example, runtimes MAY represent timestamps using a pair that tracks the iteration count of looping queries in the second component, while comparing these pairs using product order.
+
+The database state is modeled as a function of time and the program under evaluation, and all tuples are associated with the timestamp at which they were computed.
+
+PomoDB timestamps form a logical clock and hold no meaning to any other instances of PomoDB. The static nature of these links means that queries can reason from causes to effects, or from effects to their causes.
+
+## 2.3 Content Addressing
+
+As PomoDB is intended for use in distributed and decentralized deployments, it is important ensure the use of collision resistant identifiers when referring to tuples. For this purpose, content addressing scheme is used. Tuples are associated with a CID computed from their structure. The details behind this computation are available in [serialization].
+
+The choice of CIDs as primary key — rather than more familiar choices such as auto incrementing IDs or UUIDs — reflects PomoDB's goals in targeting distributed and decentralized environments, where coordination around the allocation of IDs can't be guaranteed, and where resilience against malicious and byzantine actors is required.
+
+Since content addressing schemes are backed by cryptographically secure hash functions, their use here prevents forgery of IDs by attackers, and guarantees that CID-based dependencies between tuples will always be acyclic.
+
+These properties further enable [byzantine-fault tolerant CRDTs][BFT-CRDTs].
+
+## 2.4 Query Engine
+
+PomoDB has no specified high-level query language. An intermediate representation based on datalog is defined instead ([PomoLogic]). Implementations MAY define their own user-facing query language, but they are RECOMMENDED to treat [PomoLogic] as a common compilation target for all such languages.
+
+An OPTIONAL relational runtime that MAY be compiled from [PomoLogic] is described in [PomoRA]. This can be further optimized with a dataflow runtime for incrementalization.
+
+## 2.5 Evaluation
+
+Evaluation of PomoDB queries proceeds in timesteps ("epochs") which compute a least fixed point over a batch of changes to the database.
+
+The details behind this computation are runtime specific, however all runtimes MUST provide the following additional guarantees.
 
-A n-ary relation contains n-tuples, which can each be written:
+Each epoch is denoted by the timestamp succeeding the last, and begins by scheduling the program's [sources] for evaluation. These sources MAY run in any order, however any operations over their resulting contents MUST be deferred until their completion.
```

Comment:
I'm a little iffy on this phrasing, because it makes it sound like the sources are blocking instead of asynchronous. In the current implementation, the sources constantly run asynchronously in the background, and toss incoming tuples into the runtime to be processed at the beginning of the next epoch.

This is clearer in the next section, so it's just this specific sentence that is unclear to me:
> These sources MAY run in any order, however any operations over their resulting contents MUST be deferred until their completion.